Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,358 @@
+/-
+Copyright (c) 2023 Antoine Chambetr-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambetr-Loir
+-/
+
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.LinearAlgebra.Quotient
+
+/-! # Right exactness properties of tensor product
+
+* For linear maps `f : M →ₗ[R] N`  and `g : N →ₗ[R] P`, `Exact f g` says that `range f = ker g``
+
+
+* `TensorProduct.rTensor_exact` says that one can tensor a short exact sequence on the right
+
+
+* `TensorProduct.lTensor_exact` says that one can tensor a short exact sequence on the left
+
+* For `N : Submodule R M`, `LinearMap.exact_subtype_mkQ N` says that
+  the inclusion of the submodule and the quotient map form an exact pair,
+  and `lTensor_mkQ` compute `ker (lTensor Q (N.mkQ))` and similarly for `rTensor_mkQ`
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map g g'`
+in the presence of two short exact sequences.
+
+The proofs are those of [bourbaki1989] (chap. 2, §3, n°6)
+
+
+## TODO
+
+
+* Analogue for morphisms of algebras
+
+* Treat the noncommutative case
+
+* Treat the case of modules over semirings
+(For a possible definition of an exact sequence of commutative semigroups, see
+  [Grillet-1969b], Pierre-Antoine Grillet,
+  *The tensor product of commutative semigroups*,
+  Trans. Amer. Math. Soc. 138 (1969), 281-293, doi:10.1090/S0002-9947-1969-0237688-1 .)
+
+-/
+
+section Exact
+
+variable {M N P : Type _} (f : M → N) (g : N → P)
+
+def Exact [Zero P] : Prop := ∀ y, g y = 0 ↔ y ∈ Set.range f
+
+lemma Exact.comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [Function.comp_apply, Pi.zero_apply]
+  rw [h]
+  exact Set.mem_range_self x
+
+open LinearMap
+
+variable {R : Type _} [CommSemiring R]
+variable {M N P : Type _}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+  [Module R M] [Module R N] [Module R P]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+lemma Exact.linearMap_ker_eq (hfg : Exact f g) : ker g = range f := by
+  ext y
+  exact hfg y
+
+lemma LinearMap.exact_iff : Exact f g ↔ LinearMap.ker g = LinearMap.range f := by
+  constructor
+  · intro h; apply Exact.linearMap_ker_eq; exact h
+  · intro h y; rw [← mem_ker, h]; rfl
```

Comment:
I think this is `SetLike.ext_iff.symm`

Reply:
this is almost that, but `exact…` does not work.